mbzirc_ign/src/GameLogicPlugin.cc

```diff
@@ -0,0 +1,653 @@
+                mName->Data() + "/battery/linear_battery/state";
+              this->dataPtr->node.Subscribe(batteryTopic,
+                  &GameLogicPluginPrivate::OnBatteryMsg, this->dataPtr.get());
+            }
+          }
+          return true;
+        });
+
+    // Start automatically if setup time has elapsed.
+    if (this->dataPtr->simTime.sec() >= this->dataPtr->setupTimeSec)
+    {
+      this->dataPtr->Start(this->dataPtr->simTime);
+    }
+  }
+
+  // Get the start sim time in nanoseconds.
+  auto startSimTime = std::chrono::nanoseconds(
+      this->dataPtr->startSimTime.sec() * 1000000000 +
+      this->dataPtr->startSimTime.nsec());
+
+  // Compute the elapsed competition time.
+  auto elapsedCompetitionTime = this->dataPtr->started ?
+    _info.simTime - startSimTime : std::chrono::seconds(0);
+
+  // Compute the remaining competition time.
+  auto remainingCompetitionTime = this->dataPtr->started &&
+    this->dataPtr->runDuration != std::chrono::seconds(0) ?
+    this->dataPtr->runDuration - elapsedCompetitionTime :
+    std::chrono::seconds(0) ;
+
+  // Check if the allowed time has elapsed. If so, then mark as finished.
+  if ((this->dataPtr->started && !this->dataPtr->finished) &&
+      this->dataPtr->runDuration != std::chrono::seconds(0) &&
+      remainingCompetitionTime <= std::chrono::seconds(0))
+  {
+    ignmsg << "Time limit[" <<  this->dataPtr->runDuration.count()
+      << "s] reached.\n";
+    this->dataPtr->Finish(this->dataPtr->simTime);
+  }
+
+  auto currentTime = std::chrono::steady_clock::now();
+  if (currentTime - this->dataPtr->lastStatusPubTime > std::chrono::seconds(1))
+  {
+    ignition::msgs::Clock competitionClockMsg;
+    ignition::msgs::Header::Map *mapData =
+      competitionClockMsg.mutable_header()->add_data();
+    mapData->set_key("phase");
+    if (this->dataPtr->started)
+    {
+      mapData->add_value(this->dataPtr->finished ? "finished" : "run");
+      auto secondsRemaining = std::chrono::duration_cast<std::chrono::seconds>(
+          remainingCompetitionTime);
+      competitionClockMsg.mutable_sim()->set_sec(secondsRemaining.count());
+    }
+    else if (!this->dataPtr->finished)
+    {
+      mapData->add_value("setup");
+      competitionClockMsg.mutable_sim()->set_sec(
+          this->dataPtr->setupTimeSec - this->dataPtr->simTime.sec());
+    }
+    else
+    {
+      // It's possible for a team to call Finish before starting.
+      mapData->add_value("finished");
+    }
+
+    this->dataPtr->competitionClockPub.Publish(competitionClockMsg);
+
+    this->dataPtr->lastStatusPubTime = currentTime;
+  }
+
+  // Periodically update the score file.
+  if (!this->dataPtr->finished && currentTime -
+      this->dataPtr->lastUpdateScoresTime > std::chrono::seconds(30))
+  {
+    this->dataPtr->UpdateScoreFiles(this->dataPtr->simTime);
+  }
+}
+
+/////////////////////////////////////////////////
+void GameLogicPluginPrivate::PublishScore()
```

Comment:
If the idea is to run this function in a separate thread I think we need a mutex to protect `totalScore`.

Reply:
yeah added mutex to protect `totalScore` and also started a thread to publish score. 969f88f5112dbd6d8c6894787bae404723c50a2c